app/services/payout/service.rb

```diff
@@ -0,0 +1,82 @@
+module Payout
+  class Service
+    class WalletError < StandardError; end
+
+    def initialize(payout_report:, publisher:, should_send_notifications:)
+      @publisher = publisher
+      @payout_report = payout_report
+      @should_send_notifications = should_send_notifications
+    end
+
+    def should_only_notify?
+      @payout_report.nil?
+    end
+
+    def create_message(message)
+      return if should_only_notify?
+
+      PayoutMessage.create(
+        payout_report: @payout_report,
+        publisher: @publisher,
+        message: "#{self.class.name} #{message}"
+      )
+    end
+
+    # Internal: Creates entries for any reason for not paying out a publisher.
+    #
+    # Returns a boolean
+    def skip_publisher?
+      if !@publisher.has_verified_channel?
+        create_message("Publisher has no verified channels")
+        return true
+      end
+
+      if @publisher.excluded_from_payout?
+        create_message("Publisher has been marked as excluded from payout")
+        return true
+      end
+
+      if @publisher.hold?
+        create_message("Publisher is currently in hold status")
+        return true
+      end
+
+      if @publisher.locked?
+        create_message("Publisher is currently in lock status")
+        return true
+      end
+
+      if @publisher.wire_only?
+        create_message("Publisher is only meant to receive wire transfers")
+        return true
+      end
+
+      # Checking these classes will be removed once the following issue is addressed
+      # https://github.com/brave-intl/publishers/issues/2858
+
+      if self.class == Payout::UpholdService && @publisher.gemini_connection.present?
+        # A publisher can have a gemini_connection and uphold_connection.
+        # We don't want to include users who have a GeminiConnection on the Uphold Payout
+        create_message("Publisher has a gemini connection and is not paid out through this job")
```

Comment:
Unlikely to happen given the number of Gemini signups, but it's possible that if someone adds a Gemini account after the Uphold payout ran, they'd get double paid.

Reply:
Fortunately we have validations present on the `PotentialPayment` to prevent this. https://github.com/brave-intl/publishers/blob/staging/app/models/potential_payment.rb#L13